flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java

```diff
@@ -659,128 +655,107 @@ protected int savepoint(String[] args) {
 	 */
-	private int triggerSavepoint(SavepointOptions options, JobID jobId, String savepointDirectory) {
-		try {
-			CustomCommandLine<?> activeCommandLine = getActiveCustomCommandLine(options.getCommandLine());
-			ClusterClient client = activeCommandLine.retrieveCluster(options.getCommandLine(), config, configurationDirectory);
-			try {
-				logAndSysout("Triggering savepoint for job " + jobId + ".");
-				CompletableFuture<String> savepointPathFuture = client.triggerSavepoint(jobId, savepointDirectory);
-
-				String savepointPath;
-				try {
-					logAndSysout("Waiting for response...");
-					savepointPath = savepointPathFuture.get();
-				}
-				catch (ExecutionException ee) {
-					Throwable cause = ExceptionUtils.stripExecutionException(ee);
-					throw new FlinkException("Triggering a savepoint for the job " + jobId + " failed.", cause);
-				}
-
-				logAndSysout("Savepoint completed. Path: " + savepointPath);
-				logAndSysout("You can resume your program from this savepoint with the run command.");
+	private int triggerSavepoint(ClusterClient clusterClient, JobID jobId, String savepointDirectory) throws Exception {
+		logAndSysout("Triggering savepoint for job " + jobId + ".");
+		CompletableFuture<String> savepointPathFuture = clusterClient.triggerSavepoint(jobId, savepointDirectory);
 
-				return 0;
-			}
-			finally {
-				client.shutdown();
-			}
+		String savepointPath;
+		try {
+			logAndSysout("Waiting for response...");
+			savepointPath = savepointPathFuture.get();
 		}
-		catch (Throwable t) {
-			return handleError(t);
+		catch (ExecutionException ee) {
+			Throwable cause = ExceptionUtils.stripExecutionException(ee);
+			throw new FlinkException("Triggering a savepoint for the job " + jobId + " failed.", cause);
 		}
+
+		logAndSysout("Savepoint completed. Path: " + savepointPath);
+		logAndSysout("You can resume your program from this savepoint with the run command.");
+
+		return 0;
 	}
 
 	/**
 	 * Sends a {@link org.apache.flink.runtime.messages.JobManagerMessages.DisposeSavepoint}
 	 * message to the job manager.
 	 */
-	private int disposeSavepoint(SavepointOptions options) {
-		try {
-			String savepointPath = options.getSavepointPath();
-			if (savepointPath == null) {
-				throw new IllegalArgumentException("Missing required argument: savepoint path. " +
-						"Usage: bin/flink savepoint -d <savepoint-path>");
-			}
-
-			ActorGateway jobManager = getJobManagerGateway(options);
+	private int disposeSavepoint(ClusterClient clusterClient, String savepointPath) {
+		Preconditions.checkNotNull(savepointPath, "Missing required argument: savepoint path. " +
+			"Usage: bin/flink savepoint -d <savepoint-path>");
 
-			logAndSysout("Disposing savepoint '" + savepointPath + "'.");
+		logAndSysout("Disposing savepoint '" + savepointPath + "'.");
 
-			Object msg = new DisposeSavepoint(savepointPath);
-			Future<Object> response = jobManager.ask(msg, clientTimeout);
+		CompletableFuture<Acknowledge> disposeFuture = null;
```

Comment:
nit: `null` initialization is not needed because you return in the `catch` block. Can be even declared final.

Reply:
You're right. Will change it.